packages/element/src/create-interpolate-element.js

```diff
@@ -0,0 +1,326 @@
+		throw new Error( `Not a valid element (${ element })` );
+	}
+	props.key = ++keyIndex;
+	return createElement( element, props, children );
+};
+
+/**
+ * An object describing a component to be created.
+ *
+ * This is used by the string iterator to track children that get added to an
+ * element when it is created. This allows for collecting nested elements in
+ * the string before creating the parent.
+ *
+ * @param   {Function}  creator An element creator that will be invoked for
+ *                              actually creating the react element with the
+ *                              provided children.
+ * @param   {Array}  children   Array of children to be provided to the element
+ *                              creator when invoked.
+ *
+ * @return  {Object}            An object returning the creator and children.
+ */
+function Component( creator, children = [] ) {
+	return {
+		creator,
+		children,
+	};
+}
+
+/**
+ * This encapsulates information about the current iteration state for adding to
+ * the stack
+ *
+ * @param {Component} component
+ * @param {number}    tokenStart
+ * @param {number}    tokenLength
+ * @param {number}    prevOffset
+ * @param {number}    leadingTextStart
+ *
+ * @return {Object} Iteration info as an object.
+ */
+function Frame(
+	component,
+	tokenStart,
+	tokenLength,
+	prevOffset,
+	leadingTextStart
+) {
+	return {
+		component,
+		tokenStart,
+		tokenLength,
+		prevOffset: prevOffset || tokenStart + tokenLength,
+		leadingTextStart,
+	};
+}
+
+/**
+ * This function creates an interpolated element from a passed in string with
+ * specific tags matching how the string should be converted to an element via
+ * the conversion map value.
+ *
+ * @example
+ * For example, for the given string:
+ *
+ * "This is a <span>string</span> with <a>a link</a> and a self-closing
+ * <CustomComponentB/> tag"
+ *
+ * You would have something like this as the conversionMap value:
+ *
+ * ```js
+ * {
+ *     span: ['span'],
+ *     a: ['a', { href: 'https://github.com' }],
+ *     CustomComponentB: [ CustomComponentB ],
+ * }
+ * ```
+ *
+ * @param {string}  interpolatedString  The interpolation string to be parsed.
+ * @param {Object}  conversionMap       The map used to convert the string to
+ *                                      a react element.
+ *
+ * @return {Element}  A react element.
+ */
+const createInterpolateElement = ( interpolatedString, conversionMap ) => {
+	indoc = interpolatedString;
+	offset = 0;
+	keyIndex = -1;
+	output = [];
+	stack = [];
+	tokenizer.lastIndex = 0;
+
+	if ( ! isValidConversionMap( conversionMap ) ) {
+		return interpolatedString;
+	}
+
+	do {
+		// twiddle our thumbs
+	} while ( proceed( conversionMap ) );
+
+	output = output.every( ( a ) => 'string' === typeof a ) ?
+		output.join( '' ) :
+		output.filter( ( a ) => '' !== a );
+	return typeof output === 'string' ?
+		output :
+		createElement( Fragment, {}, output );
```

Comment:
we can eliminate our second conditional…

```js
const nonEmptyChunks = output.filter( a => '' !== a );
return ! nonEmptyChunks.every( a => 'string' === typeof a )
	? createElement( Fragment, null, nonEmptyChunks )
	: nonEmptyChunks.join( '' );
```

Reply:
simplified in d980d8f